File: sdks/php/src/Module/PubSubManager.php
<?php

declare(strict_types=1);

namespace Synap\SDK\Module;

use Synap\SDK\SynapClient;

/**
 * Pub/Sub operations
 */
class PubSubManager
{
    public function __construct(
        private SynapClient $client
    ) {
    }

    /**
     * Publish a message to a topic.
     *
     * @return int Number of subscribers that received the message
     */
    public function publish(string $topic, mixed $message): int
    {
        $response = $this->client->sendCommand('pubsub.publish', [
            'topic'   => $topic,
            'payload' => $message,
        ]);

        $delivered = $response['subscribers_matched'] ?? 0;
        assert(is_int($delivered) || is_numeric($delivered));

        return (int) $delivered;
    }

    /**
     * Register a subscription on the server (HTTP transport only).
     *
     * For real-time message delivery on SynapRPC, use :meth:`observe` instead.
     *
     * @param array<string> $topics Topic patterns (supports wildcards like user.*)
     */
    public function subscribeTopics(string $subscriberId, array $topics): void
    {
        $this->client->sendCommand('pubsub.subscribe', [
            'subscriber_id' => $subscriberId,
            'topics'        => $topics,
        ]);
    }

    /**
     * Unsubscribe from topics.
     *
     * @param array<string> $topics
     */
    public function unsubscribeTopics(string $subscriberId, array $topics): void
    {
        $this->client->sendCommand('pubsub.unsubscribe', [
            'subscriber_id' => $subscriberId,
            'topics'        => $topics,
        ]);
    }

    /**
     * List active topics.
     *
     * @return array<string>
     */
    public function listTopics(): array
    {
        $response = $this->client->sendCommand('pubsub.topics', []);
        $topics = $response['topics'] ?? [];

        assert(is_array($topics));

        /** @var array<string> */
        return $topics;
    }

    /**
     * Get Pub/Sub statistics.
     *
     * @return array<string, mixed>
     */
    public function stats(): array
    {
        return $this->client->sendCommand('pubsub.stats', []);
    }

    /**
     * Subscribe and invoke callback for each push message (SynapRPC only).
     *
     * Opens a dedicated server-push TCP connection and blocks the current
     * PHP process, calling $onMessage for each received message.
     * On non-SynapRPC transports, registers the subscription via HTTP and
     * returns immediately (no real-time delivery).
     *
     * @param array<string> $topics     Topic patterns to subscribe to
     * @param callable      $onMessage  Invoked with each push-message array
     * @param callable|null $shouldStop Optional predicate; loop exits when it returns true
     */
    public function observe(
        array $topics,
        callable $onMessage,
        ?callable $shouldStop = null,
        ?string $subscriberId = null,
    ): void {
        $rpc = $this->client->getSynapRpcTransport();

        if ($rpc !== null) {
            $rpc->subscribePush($topics, $onMessage, $shouldStop);
        } else {
            // HTTP fallback — register subscription, no real-time delivery
            $sid = $subscriberId ?? ('php-sub-' . time());
            $this->subscribeTopics($sid, $topics);
        }
    }
}

File: sdks/csharp/src/Synap.SDK/Modules/PubSubManager.cs
using System.Text.Json;

namespace Synap.SDK.Modules;

/// <summary>
/// Pub/Sub operations.
/// </summary>
public sealed class PubSubManager
{
    private readonly SynapClient _client;

    internal PubSubManager(SynapClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Publishes a message to a topic.
    /// </summary>
    /// <returns>Number of subscribers that received the message.</returns>
    public async Task<long> PublishAsync(
        string topic,
        object? message,
        CancellationToken cancellationToken = default)
    {
        using var response = await _client.SendCommandAsync(
            "pubsub.publish",
            new Dictionary<string, object?> { ["topic"] = topic, ["payload"] = message },
            cancellationToken).ConfigureAwait(false);

        return response.RootElement.TryGetProperty("subscribers_matched", out var sm)
            ? sm.GetInt64()
            : 0L;
    }

    /// <summary>
    /// Registers a subscription on the server (HTTP transport only).
    /// For real-time delivery on SynapRPC, use <see cref="ObserveAsync"/>.
    /// </summary>
    public async Task SubscribeTopicsAsync(
        string subscriberId,
        IEnumerable<string> topics,
        CancellationToken cancellationToken = default)
    {
        using var _ = await _client.SendCommandAsync(
            "pubsub.subscribe",
            new Dictionary<string, object?>
            {
                ["subscriber_id"] = subscriberId,
                ["topics"]        = topics.Cast<object?>().ToArray(),
            },
            cancellationToken).ConfigureAwait(false);
    }

    /// <summary>Unsubscribes from topics.</summary>
    public async Task UnsubscribeTopicsAsync(
        string subscriberId,
        IEnumerable<string> topics,
        CancellationToken cancellationToken = default)
    {
        using var _ = await _client.SendCommandAsync(
            "pubsub.unsubscribe",
            new Dictionary<string, object?>
            {
                ["subscriber_id"] = subscriberId,
                ["topics"]        = topics.Cast<object?>().ToArray(),
            },
            cancellationToken).ConfigureAwait(false);
    }

    /// <summary>Gets Pub/Sub statistics.</summary>
    public async Task<Dictionary<string, JsonElement>> StatsAsync(CancellationToken cancellationToken = default)
    {
        using var response = await _client.SendCommandAsync("pubsub.stats", null, cancellationToken).ConfigureAwait(false);

        var result = new Dictionary<string, JsonElement>();
        foreach (var property in response.RootElement.EnumerateObject())
        {
            result[property.Name] = property.Value;
        }

        return result;
    }

    /// <summary>
    /// Subscribe and yield push messages as an async stream (SynapRPC only).
    ///
    /// On a SynapRPC transport this opens a dedicated server-push TCP connection
    /// and yields messages in real time.  On other transports this registers the
    /// subscription via HTTP and returns an empty stream (no real-time delivery).
    /// </summary>
    /// <param name="topics">Topic patterns to subscribe to.</param>
    /// <param name="cancellationToken">Token used to stop the stream.</param>
    /// <returns>Async stream of message dictionaries (keys: topic, payload, id, timestamp).</returns>
    public async IAsyncEnumerable<Dictionary<string, object?>> ObserveAsync(
        IEnumerable<string> topics,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken cancellationToken = default)
    {
        var rpc = _client.GetSynapRpcTransport();
        if (rpc is not null)
        {
            await foreach (var msg in rpc.SubscribePushAsync(topics, cancellationToken).ConfigureAwait(false))
            {
                yield return msg;
            }
        }
        else
        {
            // HTTP fallback — register subscription, no real-time delivery
            var sid = $"cs-sub-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";
            await SubscribeTopicsAsync(sid, topics, cancellationToken).ConfigureAwait(false);
            // Yield nothing — callers must poll the HTTP API
        }
    }
}
